File: main/src/com/google/refine/browsing/facets/ScatterplotFacet.java
/*

Copyright 2010, Google Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

    * Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above
copyright notice, this list of conditions and the following disclaimer
in the documentation and/or other materials provided with the
distribution.
    * Neither the name of Google Inc. nor the names of its
contributors may be used to endorse or promote products derived from
this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,           
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY           
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

*/

package com.google.refine.browsing.facets;

import java.awt.Color;
import java.awt.geom.AffineTransform;
import java.awt.geom.Point2D;
import java.awt.image.BufferedImage;
import java.awt.image.RenderedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

import javax.imageio.ImageIO;

import org.apache.commons.codec.binary.Base64;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.refine.browsing.FilteredRecords;
import com.google.refine.browsing.FilteredRows;
import com.google.refine.browsing.RecordFilter;
import com.google.refine.browsing.RowFilter;
import com.google.refine.browsing.filters.AnyRowRecordFilter;
import com.google.refine.browsing.filters.DualExpressionsNumberComparisonRowFilter;
import com.google.refine.browsing.util.ExpressionBasedRowEvaluable;
import com.google.refine.browsing.util.NumericBinIndex;
import com.google.refine.browsing.util.NumericBinRecordIndex;
import com.google.refine.browsing.util.NumericBinRowIndex;
import com.google.refine.expr.Evaluable;
import com.google.refine.expr.MetaParser;
import com.google.refine.expr.ParsingException;
import com.google.refine.model.Column;
import com.google.refine.model.Project;

public class ScatterplotFacet implements Facet {

    public static final int LIN = 0;
    public static final int LOG = 1;

    public static final int NO_ROTATION = 0;
    public static final int ROTATE_CW = 1;
    public static final int ROTATE_CCW = 2;

    /*
     * Configuration, from the client side
     */
    public static class ScatterplotFacetConfig implements FacetConfig {

        @JsonProperty("name")
        protected String name; // name of facet

        @JsonProperty(X_EXPRESSION)
        protected String expression_x; // expression to compute the x numeric value(s) per row
        @JsonProperty(Y_EXPRESSION)
        protected String expression_y; // expression to compute the y numeric value(s) per row
        @JsonProperty(X_COLUMN_NAME)
        protected String columnName_x; // column to base the x expression on, if any
        @JsonProperty(Y_COLUMN_NAME)
        protected String columnName_y; // column to base the y expression on, if any

        @JsonProperty(SIZE)
        protected int size;
        @JsonIgnore
        protected int dim_x;
        @JsonIgnore
        protected int dim_y;
        @JsonIgnore
        protected String rotation_str;
        @JsonIgnore
        protected int rotation;

        @JsonIgnore
        protected double l = 1.;
        @JsonProperty(DOT)
        protected double dot;

        @JsonIgnore
        protected String color_str = "000000";

        @JsonIgnore
        protected Color getColor() {
            return new Color(Integer.parseInt(color_str, 16));
        }

        @JsonProperty(FROM_X)
        protected double from_x; // the numeric selection for the x axis, from 0 to 1
        @JsonProperty(TO_X)
        protected double to_x;
        @JsonProperty(FROM_Y)
        protected double from_y; // the numeric selection for the y axis, from 0 to 1
        @JsonProperty(TO_Y)
        protected double to_y;

        // false if we're certain that all rows will match
        // and there isn't any filtering to do
        protected boolean isSelected() {
            return from_x > 0 || to_x < 1 || from_y > 0 || to_y < 1;
        }

        @JsonProperty(DIM_X)
        public String getDimX() {
            return dim_x == LIN ? "lin" : "log";
        }

        @JsonProperty(DIM_Y)
        public String getDimY() {
            return dim_y == LIN ? "lin" : "log";
        }

        @Override
        public ScatterplotFacet apply(Project project) {
            ScatterplotFacet facet = new ScatterplotFacet();
            facet.initializeFromConfig(this, project);
            return facet;
        }

        public static int getRotation(String rotation) {
            rotation = rotation.toLowerCase();
            if ("cw".equals(rotation) || "right".equals(rotation)) {
                return ScatterplotFacet.ROTATE_CW;
            } else if ("ccw".equals(rotation) || "left".equals(rotation)) {
                return ScatterplotFacet.ROTATE_CCW;
            } else {
                return NO_ROTATION;
            }
        }

        @Override
        public String getJsonType() {
            return "scatterplot";
        }
    }

    ScatterplotFacetConfig config;

    /*
     * Derived configuration data
     */
    protected int columnIndex_x;
    protected int columnIndex_y;
    protected Evaluable eval_x;
    protected Evaluable eval_y;
    protected String errorMessage_x;
    protected String errorMessage_y;

    protected double min_x;
    protected double max_x;
    protected double min_y;
    protected double max_y;
    protected AffineTransform t;

    protected String image;

    public static final String NAME = "name";
    public static final String IMAGE = "image";
    public static final String COLOR = "color";
    public static final String BASE_COLOR = "base_color";
    public static final String SIZE = "l";
    public static final String ROTATION = "r";
    public static final String DOT = "dot";
    public static final String DIM_X = "dim_x";
    public static final String DIM_Y = "dim_y";

    public static final String X_COLUMN_NAME = "cx";
    public static final String X_EXPRESSION = "ex";
    public static final String MIN_X = "min_x";
    public static final String MAX_X = "max_x";
    public static final String TO_X = "to_x";
    public static final String FROM_X = "from_x";
    public static final String ERROR_X = "error_x";

    public static final String Y_COLUMN_NAME = "cy";
    public static final String Y_EXPRESSION = "ey";
    public static final String MIN_Y = "min_y";
    public static final String MAX_Y = "max_y";
    public static final String TO_Y = "to_y";
    public static final String FROM_Y = "from_y";
    public static final String ERROR_Y = "error_y";

    private static final boolean IMAGE_URI = false;

    public static String EMPTY_IMAGE;

    final static Logger logger = LoggerFactory.getLogger("scatterplot_facet");

    static {
        try {
            EMPTY_IMAGE = serializeImage(new BufferedImage(1, 1, BufferedImage.TYPE_4BYTE_ABGR));
        } catch (IOException e) {
            EMPTY_IMAGE = "";
        }
    }

    @JsonProperty(NAME)
    public String getName() {
        return config.name;
    }

    @JsonProperty(X_COLUMN_NAME)
    public String getXColumnName() {
        return config.columnName_x;
    }

    @JsonProperty(X_EXPRESSION)
    public String getXExpression() {
        return config.expression_x;
    }

    @JsonProperty(Y_COLUMN_NAME)
    public String getYColumnName() {
        return config.columnName_y;
    }

    @JsonProperty(Y_EXPRESSION)
    public String getYExpression() {
        return config.expression_y;
    }

    @JsonProperty(SIZE)
    public int getSize() {
        return config.size;
    }

    @JsonProperty(DIM_X)
    public int getDimX() {
        return config.dim_x;
    }

    @JsonProperty(DIM_Y)
    public int getDimY() {
        return config.dim_y;
    }

    @JsonProperty(DOT)
    public double getDot() {
        return config.dot;
    }

    @JsonProperty(ROTATION)
    public double getRotation() {
        return config.rotation;
    }

    @JsonProperty(COLOR)
    public String getColorString() {
        return config.color_str;
    }

    @JsonProperty(IMAGE)
    @JsonInclude(Include.NON_NULL)
    public String getImage() {
        if (IMAGE_URI) {
            return image;
        }
        return null;
    }

    @JsonProperty(ERROR_X)
    @JsonInclude(Include.NON_NULL)
    public String getErrorX() {
        return errorMessage_x;
    }

    @JsonProperty(FROM_X)
    @JsonInclude(Include.NON_NULL)
    public Double getFromX() {
        if (errorMessage_x == null && !Double.isInfinite(min_x) && !Double.isInfinite(max_x)) {
            return config.from_x;
        }
        return null;
    }

    @JsonProperty(TO_X)
    @JsonInclude(Include.NON_NULL)
    public Double getToX() {
        if (errorMessage_x == null && !Double.isInfinite(min_x) && !Double.isInfinite(max_x)) {
            return config.to_x;
        }
        return null;
    }

    @JsonProperty(ERROR_Y)
    @JsonInclude(Include.NON_NULL)
    public String getErrorY() {
        return errorMessage_y;
    }

    @JsonProperty(FROM_Y)
    @JsonInclude(Include.NON_NULL)
    public Double getFromY() {
        if (errorMessage_y == null && !Double.isInfinite(min_y) && !Double.isInfinite(max_y)) {
            return config.from_y;
        }
        return null;
    }

    @JsonProperty(TO_Y)
    @JsonInclude(Include.NON_NULL)
    public Double getToY() {
        if (errorMessage_y == null && !Double.isInfinite(min_y) && !Double.isInfinite(max_y)) {
            return config.to_y;
        }
        return null;
    }

    public void initializeFromConfig(ScatterplotFacetConfig configuration, Project project) {
        config = configuration;

        t = createRotationMatrix(config.rotation, config.l);

        if (config.columnName_x.length() > 0) {
            Column x_column = project.columnModel.getColumnByName(config.columnName_x);
            if (x_column != null) {
                columnIndex_x = x_column.getCellIndex();

                NumericBinIndex index_x = ScatterplotFacet.getBinIndex(project, x_column, eval_x, config.expression_x);
                min_x = index_x.getMin();
                max_x = index_x.getMax();
            } else {
                errorMessage_x = "No column named " + config.columnName_x;
            }
        } else {
            columnIndex_x = -1;
        }

        try {
            eval_x = MetaParser.parse(config.expression_x);
        } catch (ParsingException e) {
            errorMessage_x = e.getMessage();
        }

        if (config.columnName_y.length() > 0) {
            Column y_column = project.columnModel.getColumnByName(config.columnName_y);
            if (y_column != null) {
                columnIndex_y = y_column.getCellIndex();

                NumericBinIndex index_y = ScatterplotFacet.getBinIndex(project, y_column, eval_y, config.expression_y);
                min_y = index_y.getMin();
                max_y = index_y.getMax();
            } else {
                errorMessage_y = "No column named " + config.columnName_y;
            }
        } else {
            columnIndex_y = -1;
        }

        try {
            eval_y = MetaParser.parse(config.expression_y);
        } catch (ParsingException e) {
            errorMessage_y = e.getMessage();
        }

    }

    @Override
    public RowFilter getRowFilter(Project project) {
        if (config.isSelected() &&
                eval_x != null && errorMessage_x == null &&
                eval_y != null && errorMessage_y == null) {
            return new DualExpressionsNumberComparisonRowFilter(
                    eval_x, config.columnName_x, columnIndex_x, eval_y, config.columnName_y, columnIndex_y) {

                double from_x_pixels = config.from_x * config.l;
                double to_x_pixels = config.to_x * config.l;
                double from_y_pixels = config.from_y * config.l;
                double to_y_pixels = config.to_y * config.l;

                @Override
                protected boolean checkValues(double x, double y) {
                    Point2D.Double p = new Point2D.Double(x, y);
                    p = translateCoordinates(p, min_x, max_x, min_y, max_y, config.dim_x, config.dim_y, config.l, t);
                    return p.x >= from_x_pixels && p.x <= to_x_pixels && p.y >= from_y_pixels && p.y <= to_y_pixels;
                };
            };
        } else {
            return null;
        }
    }

    @Override
    public RecordFilter getRecordFilter(Project project) {
        RowFilter rowFilter = getRowFilter(project);
        return rowFilter == null ? null : new AnyRowRecordFilter(rowFilter);
    }

    @Override
    public void computeChoices(Project project, FilteredRows filteredRows) {
        if (eval_x != null && eval_y != null && errorMessage_x == null && errorMessage_y == null) {
            Column column_x = project.columnModel.getColumnByCellIndex(columnIndex_x);
            NumericBinIndex index_x = getBinIndex(project, column_x, eval_x, config.expression_x, "row-based");

            Column column_y = project.columnModel.getColumnByCellIndex(columnIndex_y);
            NumericBinIndex index_y = getBinIndex(project, column_y, eval_y, config.expression_y, "row-based");

            retrieveDataFromBinIndices(index_x, index_y);

            if (IMAGE_URI) {
                if (index_x.isNumeric() && index_y.isNumeric()) {
                    ScatterplotDrawingRowVisitor drawer = new ScatterplotDrawingRowVisitor(
                            columnIndex_x, columnIndex_y, min_x, max_x, min_y, max_y,
                            config.size, config.dim_x, config.dim_y, config.rotation, config.dot, config.getColor());
                    filteredRows.accept(project, drawer);

                    try {
                        image = serializeImage(drawer.getImage());
                    } catch (IOException e) {
                        logger.warn("Exception caught while generating the image", e);
                    }
                } else {
                    image = EMPTY_IMAGE;
                }
            }
        }
    }

    @Override
    public void computeChoices(Project project, FilteredRecords filteredRecords) {
        if (eval_x != null && eval_y != null && errorMessage_x == null && errorMessage_y == null) {
            Column column_x = project.columnModel.getColumnByCellIndex(columnIndex_x);
            NumericBinIndex index_x = getBinIndex(project, column_x, eval_x, config.expression_x, "record-based");

            Column column_y = project.columnModel.getColumnByCellIndex(columnIndex_y);
            NumericBinIndex index_y = getBinIndex(project, column_y, eval_y, config.expression_y, "record-based");

            retrieveDataFromBinIndices(index_x, index_y);

            if (IMAGE_URI) {
                if (index_x.isNumeric() && index_y.isNumeric()) {
                    ScatterplotDrawingRowVisitor drawer = new ScatterplotDrawingRowVisitor(
                            columnIndex_x, columnIndex_y, min_x, max_x, min_y, max_y,
                            config.size, config.dim_x, config.dim_y, config.rotation, config.dot, config.getColor());
                    filteredRecords.accept(project, drawer);

                    try {
                        image = serializeImage(drawer.getImage());
                    } catch (IOException e) {
                        logger.warn("Exception caught while generating the image", e);
                    }
                } else {
                    image = EMPTY_IMAGE;
                }
            }
        }
    }

    protected void retrieveDataFromBinIndices(NumericBinIndex index_x, NumericBinIndex index_y) {
        min_x = index_x.getMin();
        max_x = index_x.getMax();

        min_y = index_y.getMin();
        max_y = index_y.getMax();
    }

    public static String serializeImage(RenderedImage image) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream(4096);
        ImageIO.write(image, "png", output);
        output.close();
        String encoded = Base64.encodeBase64String(output.toByteArray());
        String url = "data:image/png;base64," + encoded;
        return url;
    }

    public static int getAxisDim(String type) {
        return ("log".equals(type.toLowerCase())) ? LOG : LIN;
    }

    public static NumericBinIndex getBinIndex(Project project, Column column, Evaluable eval, String expression) {
        return getBinIndex(project, column, eval, expression, "row-based");
    }

    public static NumericBinIndex getBinIndex(Project project, Column column, Evaluable eval, String expression, String mode) {
        String key = "numeric-bin:" + mode + ":" + expression;
        if (eval == null) {
            try {
                eval = MetaParser.parse(expression);
            } catch (ParsingException e) {
                logger.warn("Error parsing expression", e);
            }
        }
        NumericBinIndex index = (NumericBinIndex) column.getPrecompute(key);
        if (index == null) {
            index = "row-based".equals(mode)
                    ? new NumericBinRowIndex(project, new ExpressionBasedRowEvaluable(column.getName(), column.getCellIndex(), eval))
                    : new NumericBinRecordIndex(project, new ExpressionBasedRowEvaluable(column.getName(), column.getCellIndex(), eval));

            column.setPrecompute(key, index);
        }
        return index;
    }

    private static double s_rotateScale = 1 / Math.sqrt(2.0);

    public static AffineTransform createRotationMatrix(int rotation, double l) {
        if (rotation == ScatterplotFacet.ROTATE_CW) {
            AffineTransform t = AffineTransform.getTranslateInstance(0, l / 2);
            t.scale(s_rotateScale, s_rotateScale);
            t.rotate(-Math.PI / 4);
            return t;
        } else if (rotation == ScatterplotFacet.ROTATE_CCW) {
            AffineTransform t = AffineTransform.getTranslateInstance(l / 2, 0);
            t.scale(s_rotateScale, s_rotateScale);
            t.rotate(Math.PI / 4);
            return t;
        } else {
            return null;
        }
    }

    public static Point2D.Double translateCoordinates(
            Point2D.Double p,
            double min_x, double max_x, double min_y, double max_y,
            int dim_x, int dim_y, double l, AffineTransform t) {

        double x = p.x;
        double y = p.y;

        double relative_x = x - min_x;
        double range_x = max_x - min_x;
        if (dim_x == ScatterplotFacet.LOG) {
            x = Math.log10(relative_x + 1) * l / Math.log10(range_x + 1);
        } else {
            x = relative_x * l / range_x;
        }

        double relative_y = y - min_y;
        double range_y = max_y - min_y;
        if (dim_y == ScatterplotFacet.LOG) {
            y = Math.log10(relative_y + 1) * l / Math.log10(range_y + 1);
        } else {
            y = relative_y * l / range_y;
        }

        p.x = x;
        p.y = y;
        if (t != null) {
            t.transform(p, p);
        }

        return p;
    }

}

File: openrefine-2.8/webapp/modules/core/scripts/facets/scatterplot-facet.js
/*

Copyright 2010, Google Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

 * Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above
copyright notice, this list of conditions and the following disclaimer
in the documentation and/or other materials provided with the
distribution.
 * Neither the name of Google Inc. nor the names of its
contributors may be used to endorse or promote products derived from
this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,           
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY           
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 */

function ScatterplotFacet(div, config, options) {
  this._div = div;
  this._config = config;
  this._options = options;

  this._error = false;
  this._initializedUI = false;
}

ScatterplotFacet.prototype.update = function() {
  this._plotAreaSelector.update();
};

ScatterplotFacet.prototype.reset = function() {
  delete this._config.from_x;
  delete this._config.from_y;
  delete this._config.to_x;
  delete this._config.to_y;
  this._plotAreaSelector.setOptions({ hide : true });
  this._plotAreaSelector.update();
};

ScatterplotFacet.reconstruct = function(div, uiState) {
  return new ScatterplotFacet(div, uiState.c, uiState.o);
};

ScatterplotFacet.prototype.dispose = function() {
  this._elmts.plotImg.imgAreaSelect({ hide : true });
};

ScatterplotFacet.prototype.getUIState = function() {
  var json = {
      c: this.getJSON(),
      o: this._options
  };

  return json;
};

ScatterplotFacet.prototype.getJSON = function() {
  this._config.type = "scatterplot";
  var dot = this._config.dot;
  if (typeof dot == 'number') this._config.dot.toFixed(2);
  return this._config;
};

ScatterplotFacet.prototype.hasSelection = function() {
  return  ("from_x" in this._config && this._config.from_x !== 0) ||
  ("from_y" in this._config && this._config.from_y !== 0) ||
  ("to_x" in this._config && this._config.to_x !== 1) ||
  ("to_y" in this._config && this._config.to_y !== 1);
};

ScatterplotFacet.prototype._initializeUI = function() {
  var self = this;
  var container = this._div.empty().show();

  var facet_id = container.attr("id");

  this._div.empty().show().html(
      '<div class="facet-title">' +
      '<div class="grid-layout layout-tightest layout-full"><table><tr>' +
      '<td width="1%"><a href="javascript:{}" title="'+$.i18n._('core-facets')["remove-facet"]+'" class="facet-title-remove" bind="removeButton">&nbsp;</a></td>' +
      '<td>' +
      '<a href="javascript:{}" class="facet-choice-link" bind="resetButton">'+$.i18n._('core-facets')["reset"]+'</a>' +
      '<span bind="titleSpan"></span>' +
      '</td>' +
      '</tr></table></div>' +
      '</div>' +
      '<div class="facet-scatterplot-body" bind="bodyDiv">' +
      '<div class="facet-scatterplot-message" bind="messageDiv">'+$.i18n._('core-facets')["loading"]+'</div>' +
      '<table width="100%"><tr>' + 
      '<td>' +
      '<div class="facet-scatterplot-plot-container">' +
      '<div class="facet-scatterplot-plot" bind="plotDiv">' +
      '<img class="facet-scatterplot-image" bind="plotBaseImg" />' +
      '<img class="facet-scatterplot-image" bind="plotImg" />' +
      '</div>' +
      '</div>' +
      '</td>' +
      '<td class="facet-scatterplot-selectors-container" width="100%">' +
      '<div class="scatterplot-selectors" bind="selectors">' +
      '<div class="buttonset scatterplot-dim-selector">' +
      '<input type="radio" id="' + facet_id + '-dim-lin" name="' + facet_id + '-dim" value="lin"/><label class="dim-lin-label" for="' + facet_id + '-dim-lin" title="'+$.i18n._('core-facets')["linear-plot"]+'">'+$.i18n._('core-facets')["linear-plot-abbr"]+'</label>' +
      '<input type="radio" id="' + facet_id + '-dim-log" name="' + facet_id + '-dim" value="log"/><label class="dim-log-label" for="' + facet_id + '-dim-log" title="'+$.i18n._('core-facets')["logar-plot"]+'">'+$.i18n._('core-facets')["logar-plot-abbr"]+'</label>' +
      '</div>' + 
      '<div class="buttonset scatterplot-rot-selector">' +
      '<input type="radio" id="' + facet_id + '-rot-ccw"  name="' + facet_id + '-rot" value="ccw"/><label class="rot-ccw-label" for="' + facet_id + '-rot-ccw" title="'+$.i18n._('core-facets')["rotated-counter-clock"]+'">&nbsp;</label>' +
      '<input type="radio" id="' + facet_id + '-rot-none" name="' + facet_id + '-rot" value="none"/><label class="rot-none-label" for="' + facet_id + '-rot-none" title="'+$.i18n._('core-facets')["no-rotation"]+'">&nbsp;</label>' +
      '<input type="radio" id="' + facet_id + '-rot-cw"   name="' + facet_id + '-rot" value="cw"/><label class="rot-cw-label" for="' + facet_id + '-rot-cw" title="'+$.i18n._('core-facets')["rotated-clock"]+'">&nbsp;</label>' +
      '</div>' +
      '<div class="buttonset scatterplot-dot-selector">' +
      '<input type="radio" id="' + facet_id + '-dot-small"   name="' + facet_id + '-dot" value="small"/><label class="dot-small-label" for="' + facet_id + '-dot-small" title="'+$.i18n._('core-facets')["small-dot"]+'">&nbsp;</label>' +
      '<input type="radio" id="' + facet_id + '-dot-regular" name="' + facet_id + '-dot" value="regular"/><label class="dot-regular-label" for="' + facet_id + '-dot-regular" title="'+$.i18n._('core-facets')["regular-dot"]+'">&nbsp;</label>' +
      '<input type="radio" id="' + facet_id + '-dot-big"     name="' + facet_id + '-dot" value="big"/><label class="dot-big-label" for="' + facet_id + '-dot-big" title="'+$.i18n._('core-facets')["big-dot"]+'">&nbsp;</label>' +
      '</div>' +
      '<div class="scatterplot-export-plot"><a bind="exportPlotLink" class="action" target="_blank">'+$.i18n._('core-facets')["export-plot"]+'</a></div>' +
      '</div>' +
      '</td>' + 
      '</tr></table>' +
      '<div class="facet-scatterplot-status" bind="statusDiv"></div>' +
      '</div>'
  );
  this._elmts = DOM.bind(this._div);

  this._elmts.titleSpan.text(this._config.name);
  this._elmts.removeButton.click(function() { self._remove(); });
  this._elmts.resetButton.click(function() {
    self.reset();
    self._updateRest();
  });

  this._elmts.plotDiv.width(this._config.l + "px").height(this._config.l + "px");
  this._elmts.plotBaseImg.attr("src", this._formulateBaseImageUrl())
  .attr("width", this._config.l)
  .attr("height", this._config.l);
  this._elmts.plotImg.attr("src", this._formulateCurrentImageUrl())
  .attr("width", this._config.l)
  .attr("height", this._config.l);

  var ops = {
      instance: true,        
      handles: false,
      parent: this._elmts.plotDiv,
      fadeSpeed: 70,
      onSelectEnd: function(elmt, selection) {
        self._putSelectionOptions(selection);
        self._updateRest();
      }
  };

  this._fillSelectionOptions(ops);
  this._plotAreaSelector = this._elmts.plotImg.imgAreaSelect(ops);

  if (this._config.dim_x == 'lin' && this._config.dim_y == 'lin') {
    this._elmts.selectors.find("#" + facet_id + "-dim-lin").attr('checked','checked');
  } else if (this._config.dim_x == 'log' && this._config.dim_y == 'log') {
    this._elmts.selectors.find("#" + facet_id + "-dim-log").attr('checked','checked');
  }

  if (this._config.r == 'cw') {
    this._elmts.selectors.find("#" + facet_id + "-rot-cw").attr('checked','checked');
  } else if (this._config.r == 'ccw') {
    this._elmts.selectors.find("#" + facet_id + "-rot-ccw").attr('checked','checked');
  } else {
    this._elmts.selectors.find("#" + facet_id + "-rot-none").attr('checked','checked');
  }

  if (this._config.dot >= 1.2) {
    this._elmts.selectors.find("#" + facet_id + "-dot-big").attr('checked','checked');
  } else if (this._config.dot <= 0.4) {
    this._elmts.selectors.find("#" + facet_id + "-dot-small").attr('checked','checked');
  } else {
    this._elmts.selectors.find("#" + facet_id + "-dot-regular").attr('checked','checked');
  }

  this._elmts.selectors.find(".scatterplot-dim-selector").change(function() {
    var dim = $(this).find("input:checked").val();
    self._config.dim_x = dim;
    self._config.dim_y = dim;
    self.reset();
    self._updateRest();
    self.changePlot();
  });

  this._elmts.selectors.find(".scatterplot-rot-selector").change(function() {
    self._config.r = $(this).find("input:checked").val();
    self.reset();
    self._updateRest();
    self.changePlot();        
  });

  this._elmts.selectors.find(".scatterplot-dot-selector").change(function() {
    var dot_size = $(this).find("input:checked").val();
    if (dot_size == "small") {
      self._config.dot = 0.4;
    } else if (dot_size == "big") {
      self._config.dot = 1.4;
    } else {
      self._config.dot = 0.8;
    }
    self.changePlot();
  });

  this._elmts.selectors.find(".buttonset").buttonset();
};

ScatterplotFacet.prototype._fillSelectionOptions = function(ops) {
  if (this.hasSelection()) {
    ops.x1 = this._config.l * this._config.from_x;
    ops.x2 = this._config.l * this._config.to_x;

    ops.y1 = this._config.l - (this._config.l * this._config.to_y);
    ops.y2 = this._config.l - (this._config.l * this._config.from_y);
  } else {
    ops.x1 = ops.y1 = 0;
    ops.x2 = ops.y2 = this._config.l;
    ops.hide = true;
  }
};

ScatterplotFacet.prototype._putSelectionOptions = function(selection) {
  if (selection.height === 0 || selection.width === 0) {
    this.reset();
  } else {
    this._config.from_x = selection.x1 / this._config.l;
    this._config.to_x = selection.x2 / this._config.l;

    this._config.from_y = (this._config.l - selection.y2) / this._config.l;
    this._config.to_y = (this._config.l - selection.y1) / this._config.l;
  }
};

ScatterplotFacet.prototype._formulateCurrentImageUrl = function() {
  return this._formulateImageUrl(ui.browsingEngine.getJSON(false, this), { color: "ff6a00" });
};

ScatterplotFacet.prototype._formulateBaseImageUrl = function() {
  return this._formulateImageUrl({},{ color: "888888", dot : this._config.dot * 0.9 });
};

ScatterplotFacet.prototype._formulateExportImageUrl = function() {
  return this._formulateImageUrl(ui.browsingEngine.getJSON(false, this), { dot : this._config.dot * 5, l: 500, base_color: "888888" });
};

ScatterplotFacet.prototype._formulateImageUrl = function(engineConfig, conf) {
  var options = {};
  for (var p in this._config) {
    if (this._config.hasOwnProperty(p)) {        
      options[p] = this._config[p];
    }
  }
  for (var p in conf) {
    if (conf.hasOwnProperty(p)) {        
      options[p] = conf[p];
    }
  }
  var params = {
      project: theProject.id,
      engine: JSON.stringify(engineConfig), 
      plotter: JSON.stringify(options) 
  };
  return "command/core/get-scatterplot?" + $.param(params);
};

ScatterplotFacet.prototype.updateState = function(data) {
  if ("error" in data) {
    this._error = true;
    this._errorMessage = "error" in data ? data.error : $.i18n._('core-facets')["unknown-error"]+".";
  } else {
    this._error = false;

    // These are in 0 - 1 coordinates
    if ("from_x" in data) {
      this._config.from_x = Math.min(Math.max(data.from_x, 0), 1);
    } else {
      this._config.from_x = 0;
    }
    if ("to_x" in data) {
      this._config.to_x = Math.min(Math.max(data.to_x, data.from_x), 1);
    } else {
      this._config.to_x = 1;
    }

    if ("from_y" in data) {
      this._config.from_y = Math.min(Math.max(data.from_y, 0), 1);
    } else {
      this._config.from_y = 0;
    }
    if ("to_y" in data) {
      this._config.to_y = Math.min(Math.max(data.to_y, data.from_y), this._config.l);
    } else {
      this._config.to_y = 1;
    }

    if (this._plotAreaSelector) {
      var ops = {};
      this._fillSelectionOptions(ops);
      this._plotAreaSelector.setOptions(ops);
      this._plotAreaSelector.update();
    }
  }

  this.render();
};

ScatterplotFacet.prototype.changePlot = function() {
  this._elmts.plotBaseImg.attr("src", this._formulateBaseImageUrl());
  this._elmts.plotImg.attr("src", this._formulateCurrentImageUrl());
  this._elmts.exportPlotLink.attr("href", this._formulateExportImageUrl());
};

ScatterplotFacet.prototype.render = function() {
  if (!this._initializedUI) {
    this._initializeUI();
    this._initializedUI = true;
  }

  if (this._error) {
    this._elmts.messageDiv.text(this._errorMessage).show();
    this._elmts.plotDiv.hide();
    this._elmts.statusDiv.hide();
    return;
  }

  this._elmts.messageDiv.hide();
  this._elmts.plotDiv.show();
  this._elmts.statusDiv.show();

  this._elmts.plotImg.attr("src", this._formulateCurrentImageUrl());
  this._elmts.exportPlotLink.attr("href", this._formulateExportImageUrl());
};

ScatterplotFacet.prototype._remove = function() {
  ui.browsingEngine.removeFacet(this);

  this._div = null;
  this._config = null;
};

ScatterplotFacet.prototype._updateRest = function() {
  Refine.update({ engineChanged: true });
};
